pyispyb/core/routes/webservices/userportalsync.py

```diff
@@ -0,0 +1,48 @@
+import logging
+from fastapi import HTTPException, status, Depends
+from fastapi.responses import Response, JSONResponse
+from pyispyb.app.base import AuthenticatedAPIRouter
+from ...modules import userportalsync as crud
+from ...schemas import userportalsync as schema
+from ....dependencies import permission
+
+
+logger = logging.getLogger("ispyb")
+
+router = AuthenticatedAPIRouter(prefix="/userportalsync", tags=["User Portal Sync"])
+
+
+@router.post(
+    "/sync_proposal",
+    response_model=schema.UserPortalProposalSync,
+)
+def sync_proposal(
+    proposal: schema.UserPortalProposalSync,
+    depends: bool = Depends(permission("uportal_sync")),
+):
+    """Create/Update a proposal from the User Portal and all its related entities"""
+    try:
+        execution_time = crud.sync_proposal(proposal=proposal)
+        proposal_dict = proposal.dict()
+        return JSONResponse(
+            status_code=200,
+            content={
+                "message": f"The proposal {proposal_dict['proposal']['proposalCode']}"
+                f"-{proposal_dict['proposal']['proposalNumber']} has been synchronized in {execution_time}"
+            },
+        )
+    except Exception as e:
+        logging.debug(e)
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+@router.get(
+    "/sync_proposal_json_schema",
```

Comment:
I wonder if this should be /sync_proposal/schema ?

Reply:
Might actually not need this method at all, it can be determined from the openapi spec directly?

Reply:
Yep, way better  /sync_proposal/schema. I thought the schema could be exposed to be used if needed by the external User Portal application for data validation purposes 

https://github.com/ispyb/py-ispyb/pull/210/commits/1f489e9939cd943b930f90e348f70c832eb59b6e

Reply:
In fact the schema is automatically made available by fastapi (all of them are, its how the swagger-ui is created), check out:
http://localhost:8000/ispyb/api/v1/openapi.json
components -> schemas -> UserPortalProposalSync